Provide consistency in supervisor clap cmds, correcting help menu problems

jeremymv2 opened the pull request:
Resolves https://github.com/habitat-sh/habitat/issues/5363
Resolves https://github.com/habitat-sh/habitat/issues/5377

This PR accomplishes several Items, enumerated below:

1) Moves all supervisor related clap subcommand definitions into functions
in hab::cli so that they are consistent and not duplicated in `hab-sup`
where they can be prone to drift.
2) Enumerates _all_ supervisor related clap subcommands under the `hab-sup`
clap app, including the supervisor commands only implemented in `hab`.
This is so that the _full_ supervisor help menu can be displayed for
all supervisor related commands in _both_ `hab` and `hab-sup`.
Previously, depending on which `sup * --help` command
was run, different help menus would be displayed. This was the cause of #5363 
3) Delegates `hab help sup` to `hab-sup`, passing only "help", so that the
full help menu is displayed. Previously, this command would route to `hab`,
displaying only the supervisor related commands run from `hab`
4) Defines a custom usage message for all `hab-sup` related supervisor
commands, making it clearer that the correct usage is `hab sup ..` not
`hab-sup ..`
5) Delegates `hab sup run -h` and `hab sup run --help` to `hab-sup` where
clap then correctly displays the help menu as opposed to routing to the
launcher where the launcher would actually attempt a start. This was
the cause of #5377 above.

Demonstrating the correct full help menu of all supervisor related
commands in `hab` and `hab-sup`:

```
✔ ~/habitat [jeremymv2/dry_up_sup_clap L|●5✚ 2⚑ 14]
17:28 $ sudo -E RUST_LOG=info target/debug/hab sup --help
hab-sup 0.60.0-dev

Authors: The Habitat Maintainers <humans@habitat.sh>
The Habitat Supervisor

USAGE:
    hab sup <SUBCOMMAND>

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

SUBCOMMANDS:
    bash      Start an interactive Bash-like shell
    depart    Depart a Supervisor from the gossip ring; kicking and banning the target from joining again with the
              same member-id
    help      Prints this message or the help of the given subcommand(s)
    run       Run the Habitat Supervisor
    secret    Commands relating to a Habitat Supervisor's Contorl Gateway secret
    sh        Start an interactive Bourne-like shell
    status    Query the status of Habitat services.
    term      Gracefully terminate the Habitat Supervisor and all of its running services
✔ ~/habitat [jeremymv2/dry_up_sup_clap L|●5✚ 2⚑ 14]
17:28 $
```

Signed-off-by: Jeremy J. Miller <jm@chef.io>

thesentinels commented:
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

baumanj commented:
I actually worked some something very similar to this awhile back, and ran into the problem that we don't want the supervisor to have a dependency on the `hab` binary. See https://github.com/habitat-sh/habitat/pull/4024 for the full details. If you've found an alternative approach, let me know, but I wanted to make sure we consider what we've already decided against before I start reviewing this in earnest.

jeremymv2 commented:
@baumanj I see. I would point out that there is a distinction though between this PR and #4024 in that this is one declares a `hab-sup` [dependency](https://github.com/habitat-sh/habitat/pull/5379/files#diff-db45192f88e6cbf89ed71247235b97eeR26) on `hab`, which is more innocuous in my opinion.  Whereas in the other PR, it was the other way around, a dependency of `hab` on `hab-sup`.

baumanj commented:
That is a good point. And at some point we're looking to possibly combine the `hab` and `hab-sup` binaries and this may become a non-issue. In the meantime, I'd want the approval of at least @christophermaier and @fnichol before we introduce a `hab-sup` → `hab` dependency.

What I'd really like to see is a switch to using external YAML files to define our CLI. CLAP supports this and it should resolve the dependency issues as well as pave the way for localization of our CLI.

christophermaier commented on components/hab/src/cli.rs:
```diff
@@ -546,6 +547,15 @@ pub fn get() -> App<'static, 'static> {
             )
         )
+        (@subcommand sup =>
+            (@setting VersionlessSubcommands)
```
I think we need an `(about: ... )` line here... my output looks like this:
``` sh
# snip
    plan             Commands relating to plans and other app-specific configuration.
    ring             Commands relating to Habitat rings
    studio           Commands relating to Habitat Studios
    sup              
    supportbundle    Create a tarball of Habitat Supervisor data to send to support
    svc              Commands relating to Habitat services
    user             Commands relating to Habitat users
# snip
```

baumanj commented:
I'm just about at the end of my workday, but now that @christophermaier is on-board, I'll give it a thorough review, Monday. Thanks again for all your help, @jeremymv2 !

baumanj commented:
Thought I would get to this today, but apparently not. Hopefully tomorrow.

baumanj commented on components/hab/src/cli.rs:
```diff
@@ -546,6 +547,15 @@ pub fn get() -> App<'static, 'static> {
+            (@setting SubcommandRequiredElseHelp)
+            // these are the only sup cmds handled in `hab`, the remaining
+            // supervisor related commands are handled in `hab-sup`
+            (subcommand: sub_sup_depart().aliases(&["d", "de", "dep", "depa", "depart"]))
```
This command is supposed to be undocumented, but with these changes it ends up displayed on `hab sup` when it wasn't before.

```
➤ env RUST_LOG= HAB_SUP_BINARY=(pwd)/target/debug/hab-sup target/debug/hab sup
hab-sup 0.60.0-dev

Authors: The Habitat Maintainers <humans@habitat.sh>
The Habitat Supervisor

USAGE:
    hab sup <SUBCOMMAND>

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

SUBCOMMANDS:
    bash      Start an interactive Bash-like shell
    depart    Depart a Supervisor from the gossip ring; kicking and banning the target from joining again with the
              same member-id
    help      Prints this message or the help of the given subcommand(s)
    run       Run the Habitat Supervisor
    secret    Commands relating to a Habitat Supervisor's Contorl Gateway secret
    sh        Start an interactive Bourne-like shell
    status    Query the status of Habitat services.
    term      Gracefully terminate the Habitat Supervisor and all of its running services
```
vs
```
➤ env RUST_LOG= hab sup
hab-sup 0.59.0/20180712161546

Authors: The Habitat Maintainers <humans@habitat.sh>
The Habitat Supervisor

USAGE:
    hab-sup <SUBCOMMAND>

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

SUBCOMMANDS:
    bash    Start an interactive Bash-like shell
    help    Prints this message or the help of the given subcommand(s)
    run     Run the Habitat Supervisor
    sh      Start an interactive Bourne-like shell
    term    Gracefully terminate the Habitat Supervisor and all of its running services
```

christophermaier replied:
This actually _should_ be documented... it's a frequently-asked question. The fact that it wasn't before was, I think, an unintended side-effect of how our CLI is currently split up.

baumanj commented on components/hab/src/cli.rs:
```diff
@@ -546,6 +547,15 @@ pub fn get() -> App<'static, 'static> {
+            // these are the only sup cmds handled in `hab`, the remaining
+            // supervisor related commands are handled in `hab-sup`
+            (subcommand: sub_sup_depart().aliases(&["d", "de", "dep", "depa", "depart"]))
+            (subcommand: sub_sup_secret().aliases(&["sec", "secr"]))
```
Similar to `depart`, `secret` was previously undocumented, but I'm not sure if that was intention or not. @christophermaier, what do you think?

christophermaier replied:
Yup, `secret` is supposed to be documented, too; see #5363

baumanj commented:
based on @christophermaier's comments, I've changed my review to approval. Once you address 
https://github.com/habitat-sh/habitat/pull/5379#pullrequestreview-141247119, you're good to go as far as I'm concerned!

jeremymv2 commented:
Please accept my apologies @christophermaier and @baumanj, I've not disappeared off the face of the earth without just cause. Our team was hit with a large amount of customer facing work this week. Then my household came down with some sort of virus funk.

I thank you for your time spent reviewing this PR and I will address the feedback item on Monday next week!

jeremymv2 commented:
@christophermaier @baumanj I have addressed the feedback from https://github.com/habitat-sh/habitat/pull/5379#pullrequestreview-141247119 and rebased on current master.
